### PR TITLE
[ENT-227] Add support for Django 1.10

### PR DESCRIPTION
**Pull request by snowcrshd**
**Description:** Adds support to Django 1.10 to DCF.
As of https://github.com/edx/edx-enterprise/pull/44, support for Django 1.10 on `edx-enterprise` has been dropped.

**JIRA:** https://openedx.atlassian.net/browse/ENT-227

**Dependencies:** 

**Merge deadline:** 

**Installation instructions:** 

**Testing instructions:**

`make test-all`

**Merge checklist:**

- [x] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [ ] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [ ] ~Documentation updated (not only docstrings)~
- [ ] Commits are (reasonably) squashed
- [ ] ~Translations are updated~
- [x] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:** Adding support for Django 1.11 and using a more recent version of DRF introduced **lots** of required changes. I believe a better approach would be to split this migration into several steps, instead of a single PR.

EDIT: changed description to use edx-enterprise template.

### CODE REVIEW

**Comment by snowcrshd** [comment]
@doctoryes should HEAD be passing the quality checks?

I noticed I'm getting a failure on travis due to errors in pylint.

**Comment by snowcrshd** [comment]
@doctoryes ok, so I finally got all of the tests to pass for Django 1.8, 1.9, and 1.10.

I'll send a new PR, pinning down `pylint` version to 1.7.1, then fix the pylint complaints.
After that is done with, I will rebase this PR on top of it.

What do you think?

PS: I noticed a "bug" while testing this: If I setup a virtualenv inside of the project's root folder, py.test will try to run tests inside it. This causes errors when running tox.

A way to correct this would be to point py.test specifically at the `config_models` package.

I can apply this quick fix on a new PR later.

File: requirements/base.in [hunk 1]
```diff
@@ -1,4 +1,4 @@
 # Core requirements for using this application
 
-Django>=1.8,<1.10                    # Web application framework
-djangorestframework>=3.2.3,<3.3.0
+Django>=1.8,<1.11                    # Web application framework
+djangorestframework>=3.4.0,<3.5.0
```
Comment (e-kolpakov) [inline requirements/base.in]: @snowcrshd DRF 3.6 already exists - might make sense to try >=3.6,<3.7

Comment (snowcrshd) [inline requirements/base.in]: @e-kolpakov good point. I'll have a look at their dependencies.

**Comment by doctoryes** [comment]
@snowcrshd Looks good - thanks! I assume that most of the changes were generated by a pip-compile run, correct? And - now that Django 1.11 is out, could you be convinced to add that compatibility as well? Or are there reasons why you're stopping at 1.10 for now?

**Comment by snowcrshd** [comment]
@doctoryes I was just playing it safe.

I'll have a look at the other edx-enterprise dependencies.
If they all support Django 1.11, I'll update the PR with support to the most recent version :D

**Comment by doctoryes** [comment]
@snowcrshd Your work towards Django 1.11 would be greatly appreciated - but not a necessity. We're close to starting a project to get edx-platform Django 1.11-compatible soon and it would help the effort along.

No need to change those to separate commits - they should probably be in the same commit anyway.

**Comment by snowcrshd** [comment]
This PR depends on #11 to pass the quality checks.

Any chance we could get that one merged so I could rebase this ASAP?

File: CHANGELOG.rst [hunk 2]
```diff
@@ -14,6 +14,14 @@ Change Log
...
 ~~~~~~~~~~
 
+[0.1.6] - 2017-06-02
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Added
+_____
+
+* Support for Django 1.10
+
```
Comment (e-kolpakov) [inline CHANGELOG.rst]: @snowcrshd Since you've restored a good practice of adding changelog records, could you please also cover 1.05 - 1.0.1 versions? Don't spend too much time on it - something like

    [0.1.1] - [0.1.5] - 2017-06-01
    ~~~~~~~~~~~~~~~~~~~~~
    * Commit message 1
    * Commit message 2
    * Commit message 3


@doctoryes FYI - changelog records were missing

**Comment by snowcrshd** [comment]
@e-kolpakov updated the PR with a commit adding the missing CHANGELOG entries.

**Comment by e-kolpakov** [comment]
@snowcrshd Ok, I've casted my approval already, so plesae proceed to upstream review.

**Comment by snowcrshd** [comment]
@e-kolpakov thank you!

@doctoryes would you mind having a look at these changes?

If there's anything you believe should be altered, please let me know.
I'm waiting on these to be merged and a new version published to Pypi,
so we can then update https://github.com/edx/edx-enterprise/pull/119 to use the new DCM version.

**Comment by snowcrshd** [comment]
@e-kolpakov you might want to have a look at this PR again.

**tl;dr**: ran into bug with DRF and edx-enterprise. Fixed by using a newer version of DRF. Updated this PR to reflect newer DRF version.

I was testing `edx-enterprise` and stumbled on some (possibly) bug when using DRF 3.4.7 (which as restricted due to DCM requiring DRF `>=3.4,<3.5`).

Apparently, the buggy behavior I ran into on `edx-enterprise` was fixed in DRF 3.5.0. I'm going through their changes for 3.5.0 now, to figure out exactly what changed there. They did not provide release notes for this version.

I tested up to DRF `3.6.3` (latest release ATM) and `edx-enterprise` works fine.

So I loosened the requirements in this PR, and ended having to apply a minor change to `AutoConfigModelSerializer` due to deprecated behavior in DRF `3.3.0` that is now enforced since `3.5.0`.

**Comment by doctoryes** [comment]
👍 LGTM

File: config_models/admin.py [hunk 3]
```diff
@@ -128,7 +128,7 @@ def has_output(self):
         """ Should this filter be shown? """
         return True
 
-    def choices(self, cl):
+    def choices(self, cl):  # pylint: disable=arguments-differ
```
Comment (e-kolpakov) [inline config_models/admin.py]: @snowcrshd You're so thoroughly ignoring my suggestions to make the interface consistent instead of suppressing an error :) You could at least say "no, to me unused argument is worse than mismatching signature, so I'll keep it" and I wouldn't raise this question again.

Comment (snowcrshd) [inline config_models/admin.py]: @e-kolpakov not ignoring at all. Sorry about this one.

I thought you only meant this on the `edx-enterprise` repository, which is where you wrote that comment.

I'll make the interface consistent.

**Comment by snowcrshd** [comment]
@e-kolpakov there. Removed the pylint suppression.

Check commit [2c38901d16be004945f63a5eaa3e364c83fc3f47](https://github.com/edx/django-config-models/pull/8/commits/2c38901d16be004945f63a5eaa3e364c83fc3f47)

**Comment by itsjeyd** [comment]
@snowcrshd FYI: You got conditional thumbs-up from @e-kolpakov, and you addressed the one remaining comment that he had. From the looks of it, addressing the comment didn't require fundamental/large-scale changes, so it's OK to move forward with this.

@bradenmacdonald Do we need a second thumbs-up from upstream to proceed?

**Comment by itsjeyd** [comment]
@bradenmacdonald Never mind -- asked the Enterprise team about that at today's standup, and they said we're good to merge this :)

@doctoryes @bradenmacdonald I don't have write access to this repo, so I can't flip the switch myself. Could one of you please do the honors?

CC @snowcrshd @brittneyexline @douglashall

**Comment by doctoryes** [comment]
Merged. Thanks, @snowcrshd .

**Comment by snowcrshd** [comment]
@doctoryes thanks! any chance we might get a tag and PyPi release for 0.1.6?

https://github.com/edx/edx-enterprise/pull/119 relies on having 0.1.6 published.

**Comment by doctoryes** [comment]
@snowcrshd I pushed a 0.1.6 tag - should be built/uploaded to PyPI in this build:
https://travis-ci.org/edx/django-config-models/builds/241240454

**Comment by snowcrshd** [comment]
@doctoryes awesome!
